security/pkg/stsservice/tokenmanagers/google/tokenmanager.go

```diff
@@ -0,0 +1,303 @@
+// {
+//  "Delegates": [],
+//  "Scope": [
+//      https://www.googleapis.com/auth/cloud-platform
+//  ],
+// }
+func (tm *TokenManager) constructGenerateAccessTokenRequest(fResp *federatedTokenResponse) *http.Request {
+	query := accessTokenRequest{}
+	query.Scope = append(query.Scope, scope)
+
+	jsonQuery, _ := json.Marshal(query)
+	endpoint := fmt.Sprintf(accessTokenEndpoint, tm.gCPProjectNumber)
+	req, _ := http.NewRequest("POST", endpoint, bytes.NewBuffer(jsonQuery))
+	req.Header.Add("Content-Type", contentType)
+	req.Header.Add("Authorization", "Bearer "+fResp.AccessToken)
+	reqDump, _ := httputil.DumpRequest(req, true)
+	tokenManagerLog.Debugf("Prepared access token request: \n%s", string(reqDump))
+	return req
+}
+
+func (tm *TokenManager) fetchAccessToken(federatedToken *federatedTokenResponse) (*accessTokenResponse, error) {
+	respData := &accessTokenResponse{}
+
+	req := tm.constructGenerateAccessTokenRequest(federatedToken)
+	resp, err := tm.sendRequestWithRetry(req)
+	if err != nil {
+		respCode := 0
+		if resp != nil {
+			respCode = resp.StatusCode
+		}
+		tokenManagerLog.Errorf("failed to exchange access token (HTTP status %d): %v", respCode, err)
+		return respData, fmt.Errorf("failed to exchange access token (HTTP status %d): %v", respCode, err)
+	}
+	defer resp.Body.Close()
+	respDump, _ := httputil.DumpResponse(resp, true)
+	tokenManagerLog.Debugf("Received access token response: \n%s", string(respDump))
+
+	body, _ := ioutil.ReadAll(resp.Body)
+	if err := json.Unmarshal(body, respData); err != nil {
+		tokenManagerLog.Errorf("Failed to unmarshal access token response data: %v", err)
+		return respData, fmt.Errorf("failed to unmarshal access token response data: %v", err)
+	}
+	if respData.AccessToken == "" {
+		tokenManagerLog.Errora("access token response does not have access token", string(body))
+		return respData, errors.New("access token response does not have access token. " + string(body))
+	}
+	tokenManagerLog.Debug("successfully exchanged an access token")
+	tokenReceivedTime := time.Now()
+	expireTime, _ := ptypes.Duration(&respData.ExpireTime)
+	tm.tokens.Store(tokenReceivedTime.String(), stsservice.TokenInfo{
+		TokenType:  accessToken,
+		IssueTime:  tokenReceivedTime.String(),
+		ExpireTime: tokenReceivedTime.Add(expireTime).String()})
+	return respData, nil
+}
+
+// generateSTSResp takes accessTokenResponse and generates StsResponseParameters in JSON.
+func (tm *TokenManager) generateSTSResp(atResp *accessTokenResponse) ([]byte, error) {
+	expireTime, _ := ptypes.Duration(&atResp.ExpireTime)
+	stsRespParam := stsservice.StsResponseParameters{
+		AccessToken:     atResp.AccessToken,
+		IssuedTokenType: tokenType,
+		TokenType:       "Bearer",
+		ExpiresIn:       int64(expireTime.Seconds()),
+	}
+	statusJSON, err := json.MarshalIndent(stsRespParam, "", " ")
+	return statusJSON, err
+}
+
+// DumpTokenStatus dumps all token status in JSON
+func (tm *TokenManager) DumpTokenStatus() ([]byte, error) {
```

Comment:
Where is this function called?

Reply:
It is called here https://github.com/istio/istio/blob/c167a42dc7defa7c26d9927eaa18abbde160751a/security/pkg/stsservice/server/server.go#L196